File: python/tests/integration/_async/test_agent_integration.py
"""Integration tests for Agent functionality."""

import asyncio
import os

import pytest
import pytest_asyncio

from agentbay import AsyncAgentBay
from agentbay import get_logger
from agentbay import CreateSessionParams
from agentbay  import AgentOptions

from dotenv import load_dotenv

logger = get_logger("agentbay-integration-test")
load_dotenv()


@pytest_asyncio.fixture(scope="module")
async def agent_bay():
    """Create an AsyncAgentBay instance."""
    api_key = os.getenv("AGENTBAY_API_KEY")
    if not api_key:
        pytest.skip("AGENTBAY_API_KEY environment variable not set")
    return AsyncAgentBay(api_key=api_key)


@pytest_asyncio.fixture(scope="module")
async def computer_agent_session(agent_bay):
    """Create a session for agent testing."""
    # Ensure a delay to avoid session creation conflicts
    await asyncio.sleep(3)
    params = CreateSessionParams(
        image_id="windows_latest",
    )
    session_result = await agent_bay.create(params)
    if not session_result.success or not session_result.session:
        pytest.skip("Failed to create session")

    session = session_result.session
    yield session

    # Clean up session
    try:
        await session.delete()
    except Exception as e:
        logger.info(f"Warning: Error deleting session: {e}")


@pytest_asyncio.fixture(scope="module")
async def browser_agent_session(agent_bay):
    """Create a session for agent testing."""
    # Ensure a delay to avoid session creation conflicts
    await asyncio.sleep(3)
    params = CreateSessionParams(
        image_id="imgc-0aae4rgkx5fdch0et",
    )
    session_result = await agent_bay.create(params)
    if not session_result.success or not session_result.session:
        pytest.skip("Failed to create session")

    session = session_result.session
    yield session

    # Clean up session
    try:
        await session.delete()
    except Exception as e:
        logger.info(f"Warning: Error deleting session: {e}")


@pytest_asyncio.fixture(scope="module")
async def mobile_agent_session(agent_bay):
    """Create a session for mobile agent testing."""
    # Ensure a delay to avoid session creation conflicts
    await asyncio.sleep(3)
    params = CreateSessionParams(
        image_id="mobile_latest",
    )
    session_result = await agent_bay.create(params)
    if not session_result.success or not session_result.session:
        pytest.skip("Failed to create session")

    session = session_result.session
    yield session

    # Clean up session
    try:
        await session.delete()
    except Exception as e:
        logger.info(f"Warning: Error deleting session: {e}")


@pytest.mark.asyncio
async def test_computer_execute_task_and_wait_success(computer_agent_session):
    """Test executing a flux task successfully."""
    agent = computer_agent_session.agent

    task = "create a folder named 'agentbay' in C:\\Window\\Temp"
    max_try_times = os.environ.get("AGENT_TASK_TIMEOUT")
    if not max_try_times:
        max_try_times = 100
    logger.info("🚀 task of creating folders")
    result = await agent.computer.execute_task_and_wait(task, int(max_try_times))
    assert result.success
    assert result.request_id != ""
    assert result.error_message == ""
    logger.info(f"✅ result {result.task_result}")


@pytest.mark.asyncio
async def test_computer_execute_task_success(computer_agent_session):
    """Test executing a flux task successfully."""
    agent = computer_agent_session.agent

    task = "create a folder named 'agentbay' in C:\\Window\\Temp"
    max_try_times = os.environ.get("AGENT_TASK_TIMEOUT")
    if not max_try_times:
        max_try_times = 100
    logger.info("🚀 async task of creating folders")
    result = await agent.computer.execute_task(task)
    assert result.success
    assert result.request_id != ""
    assert result.error_message == ""
    retry_times: int = 0
    query_result = None
    while retry_times < int(max_try_times):
        query_result = await agent.computer.get_task_status(result.task_id)
        assert result.success
        logger.info(
            f"⏳ Task {query_result.task_id} running 🚀: {query_result.task_action}."
        )
        if query_result.task_status == "completed":
            break
        retry_times += 1
        await asyncio.sleep(3)
    # Verify the final task status
    assert retry_times < int(max_try_times)
    logger.info(f"✅ result {query_result.task_product}")


@pytest.mark.asyncio
async def test_browser_execute_task_and_wait_success(browser_agent_session):
    """Test executing a flux task successfully."""
    agent = browser_agent_session.agent

    task = "Navigate to baidu.com and Query the date when Alibaba listed in the U.S"
    max_try_times = os.environ.get("AGENT_TASK_TIMEOUT")
    if not max_try_times:
        max_try_times = 100
    logger.info("🚀 task of Query the date when Alibaba listed in the U.S")
    options: AgentOptions = AgentOptions(use_vision=False, output_schema="text")
    result = await agent.browser.initialize(options)
    assert result.success
    result = await agent.browser.execute_task_and_wait(task, int(max_try_times))
    assert result.success
    assert result.request_id != ""
    assert result.error_message == ""
    logger.info(f"✅ result {result.task_result}")


@pytest.mark.asyncio
async def test_browser_execute_task_success(browser_agent_session):
    """Test executing a flux task successfully."""
    agent = browser_agent_session.agent

    task = "Navigate to baidu.com and Query the weather in Shanghai"
    max_try_times = os.environ.get("AGENT_TASK_TIMEOUT")
    if not max_try_times:
        max_try_times = 100
    logger.info("🚀 async task Query the weather in Shanghai.")
    result = await agent.browser.execute_task(task)
    assert result.success
    assert result.request_id != ""
    assert result.error_message == ""
    retry_times: int = 0
    query_result = None
    while retry_times < int(max_try_times):
        query_result = await agent.browser.get_task_status(result.task_id)
        assert result.success
        logger.info(
            f"⏳ Task {query_result.task_id} running 🚀: {query_result.task_action}."
        )
        if query_result.task_status == "completed":
            break
        retry_times += 1
        await asyncio.sleep(3)
    # Verify the final task status
    assert retry_times < int(max_try_times)
    logger.info(f"✅ result {query_result.task_product}")


@pytest.mark.asyncio
async def test_mobile_execute_task_and_wait_success(mobile_agent_session):
    """Test executing a mobile task successfully."""
    agent = mobile_agent_session.agent

    task = "Open WeChat app"
    max_poll_times = os.environ.get("AGENT_TASK_TIMEOUT")
    if not max_poll_times:
        max_poll_times = 100
    logger.info("🚀 task of opening WeChat app")
    result = await agent.mobile.execute_task_and_wait(
        task, max_steps=50, max_try_times=3, max_poll_times=int(max_poll_times)
    )
    assert result.success
    assert result.request_id != ""
    assert result.error_message == ""
    logger.info(f"✅ result {result.task_result}")


@pytest.mark.asyncio
async def test_mobile_execute_task_success(mobile_agent_session):
    """Test executing a mobile task successfully."""
    agent = mobile_agent_session.agent

    task = "Open WeChat app"
    max_try_times = os.environ.get("AGENT_TASK_TIMEOUT")
    if not max_try_times:
        max_try_times = 100
    logger.info("🚀 async task of opening WeChat app")
    result = await agent.mobile.execute_task(
        task, max_steps=50, max_try_times=3
    )
    assert result.success
    assert result.request_id != ""
    assert result.error_message == ""
    retry_times: int = 0
    query_result = None
    while retry_times < int(max_try_times):
        query_result = await agent.mobile.get_task_status(result.task_id)
        assert result.success
        logger.info(
            f"⏳ Task {query_result.task_id} running 🚀: {query_result.task_action}."
        )
        if query_result.task_status == "completed":
            break
        retry_times += 1
        await asyncio.sleep(3)
    # Verify the final task status
    assert retry_times < int(max_try_times)
    logger.info(f"✅ result {query_result.task_product}")

File: java/agentbay/src/test/java/com/aliyun/agentbay/test/TestAgentIntegration.java
package com.aliyun.agentbay.test;

import com.aliyun.agentbay.AgentBay;
import com.aliyun.agentbay.agent.Agent;
import com.aliyun.agentbay.exception.AgentBayException;
import com.aliyun.agentbay.model.ExecutionResult;
import com.aliyun.agentbay.model.QueryResult;
import com.aliyun.agentbay.model.SessionResult;
import com.aliyun.agentbay.session.CreateSessionParams;
import com.aliyun.agentbay.session.Session;
import com.fasterxml.jackson.annotation.JsonProperty;

import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;
import org.junit.experimental.runners.Enclosed;
import org.junit.runner.RunWith;

import static org.junit.Assert.*;

/**
 * Integration tests for Agent functionality in AgentBay Java SDK.
 * This test class is equivalent to test_agent_integration.py in Python SDK.
 */
@RunWith(Enclosed.class)
public class TestAgentIntegration {
    private static AgentBay agentBay;
    private static Session session;
    private static Agent agent;
    public static TestComputerAgent computer;
    public static TestBrowserAgent browser;

    public static class OutputSchema {
      @JsonProperty(value = "listedDate", required = true)
      private String listedDate;

      public String getListedDate() {
        return listedDate;
      }

      public void setListedDate(String listedDate) {
        this.listedDate = listedDate;
      }
    }

    /**
     * Get API key for testing
     */
    private static String getTestApiKey() {
        String apiKey = System.getenv("AGENTBAY_API_KEY");
        if (apiKey == null || apiKey.trim().isEmpty()) {
            apiKey = "akm-xxx"; // Replace with your actual API key for testing
            System.out.println("Warning: Using default API key. Set AGENTBAY_API_KEY environment variable for production use.");
        }
        return apiKey;
    }

    public static class TestComputerAgent {
      /**
       * Set up the test environment by creating a session and initializing
       * agent.
       */
      @BeforeClass
      public static void setUp() 
          throws AgentBayException, InterruptedException {
        // Ensure a delay to avoid session creation conflicts
        Thread.sleep(3000);
        String apiKey = getTestApiKey();
        System.out.println("----------------apiKey: " + apiKey);
        agentBay = new AgentBay(apiKey);

        // Create a session with windows_latest image
        CreateSessionParams params = new CreateSessionParams();
        params.setImageId("windows_latest");

        System.out.println("Creating a new session for agent testing...");
        SessionResult sessionResult = agentBay.create(params);

        if (!sessionResult.isSuccess() || sessionResult.getSession() == null) {
          throw new AgentBayException("Failed to create session: " +
                                      sessionResult.getErrorMessage());
        }

        session = sessionResult.getSession();
        agent = session.getAgent();

        System.out.println("Session created with ID: " +
                           session.getSessionId());
      }

      /**
       * Clean up resources after all tests.
       */
      @AfterClass
      public static void tearDown() {
        System.out.println("Cleaning up: Deleting the session...");
        try {
          if (session != null) {
            agentBay.delete(session, false);
          }
        } catch (Exception e) {
          System.err.println("Warning: Error deleting session: " +
                             e.getMessage());
        }
      }

      @Test
      public void testExecuteTaskSuccess() {
        System.out.println("🚀 Test: Execute task synchronously");

        String task = "create a folder named 'agentbay' in C:\\Windows\\Temp";
        String timeoutEnv = System.getenv("AGENT_TASK_TIMEOUT");
        int timeout = (timeoutEnv != null && !timeoutEnv.isEmpty())
                          ? Integer.parseInt(timeoutEnv)
                          : 300;

        ExecutionResult result =
            agent.getComputer().executeTaskAndWait(task, timeout);

        assertTrue("Task execution should succeed", result.isSuccess());
        assertNotNull("Request ID should not be null", result.getRequestId());
        assertFalse("Request ID should not be empty",
                    result.getRequestId().isEmpty());
        assertEquals("Error message should be empty", "",
                     result.getErrorMessage());
        assertNotNull("Task result should not be null", result.getTaskResult());

        System.out.println("✅ Task result: " + result.getTaskResult());
      }

      @Test
      public void testAsyncExecuteTaskSuccess() throws InterruptedException {
        System.out.println("🚀 Test: Execute task asynchronously");

        String task = "create a folder named 'agentbay' in C:\\Windows\\Temp";
        String timeoutEnv = System.getenv("AGENT_TASK_TIMEOUT");
        int timeout = (timeoutEnv != null && !timeoutEnv.isEmpty())
                          ? Integer.parseInt(timeoutEnv)
                          : 300;

        // Start async task execution
        ExecutionResult result = agent.getComputer().executeTask(task);

        assertTrue("Async task execution should succeed", result.isSuccess());
        assertNotNull("Request ID should not be null", result.getRequestId());
        assertFalse("Request ID should not be empty",
                    result.getRequestId().isEmpty());
        assertEquals("Error message should be empty", "",
                     result.getErrorMessage());
        assertNotNull("Task ID should not be null", result.getTaskId());
        assertFalse("Task ID should not be empty",
                    result.getTaskId().isEmpty());

        // Poll task status until finished
        int pollInterval = 3;
        int maxPollAttempts = timeout / pollInterval;
        int retryTimes = 0;
        QueryResult queryResult = null;

        while (retryTimes < maxPollAttempts) {
          queryResult = agent.getComputer().getTaskStatus(result.getTaskId());

          assertTrue("Query should succeed", queryResult.isSuccess());
          System.out.println("⏳ Task " + queryResult.getTaskId() +
                             " running 🚀: " + queryResult.getTaskAction());

          if ("finished".equals(queryResult.getTaskStatus())) {
            break;
          }

          retryTimes++;
          Thread.sleep(3000);
        }

        // Verify the final task status
        assertNotNull("Query result should not be null", queryResult);
        assertTrue("Task should finish within timeout",
                   retryTimes < maxPollAttempts);
        assertEquals("Task status should be finished", "finished",
                     queryResult.getTaskStatus());

        System.out.println("✅ Task result: " + queryResult.getTaskProduct());
      }
    }
public static class TestBrowserAgent {
      /**
       * Set up the test environment by creating a session and initializing
       * agent.
       */
      @BeforeClass
      public static void setUp()
          throws AgentBayException, InterruptedException {
        // Ensure a delay to avoid session creation conflicts
        Thread.sleep(3000);
        String apiKey = getTestApiKey();
        agentBay = new AgentBay(apiKey);

        // Create a session with windows_latest image
        CreateSessionParams params = new CreateSessionParams();
        params.setImageId("browser_latest");

        System.out.println("Creating a new session for agent testing...");
        SessionResult sessionResult = agentBay.create(params);

        if (!sessionResult.isSuccess() || sessionResult.getSession() == null) {
          throw new AgentBayException("Failed to create session: " +
                                      sessionResult.getErrorMessage());
        }

        session = sessionResult.getSession();
        agent = session.getAgent();
        

        System.out.println("Session created with ID: " +
                           session.getSessionId());
      }

      /**
       * Clean up resources after all tests.
       */
      @AfterClass
      public static void tearDown() {
        System.out.println("Cleaning up: Deleting the session...");
        try {
          if (session != null) {
            agentBay.delete(session, false);
          }
        } catch (Exception e) {
          System.err.println("Warning: Error deleting session: " +
                             e.getMessage());
        }
      }

      @Test
      public void testExecuteTaskSuccess() {
        System.out.println("🚀 Test: Execute task synchronously");

        String task = "查询阿里巴巴上市日期";
        String timeoutEnv = System.getenv("AGENT_TASK_TIMEOUT");
        int timeout = (timeoutEnv != null && !timeoutEnv.isEmpty())
                          ? Integer.parseInt(timeoutEnv)
                          : 300;

        ExecutionResult result =
            agent.getBrowser().executeTaskAndWait(task, timeout, false, OutputSchema.class);

        assertTrue("Task execution should succeed", result.isSuccess());
        assertNotNull("Request ID should not be null", result.getRequestId());
        assertFalse("Request ID should not be empty",
                    result.getRequestId().isEmpty());
        assertEquals("Error message should be empty", "",
                     result.getErrorMessage());
        assertNotNull("Task result should not be null", result.getTaskResult());

        System.out.println("✅ Task result: " + result.getTaskResult());
      }

      @Test
      public void testAsyncExecuteTaskSuccess() throws InterruptedException {
        System.out.println("🚀 Test: Execute task asynchronously");

        String task = "查询阿里巴巴上市日期";
        String timeoutEnv = System.getenv("AGENT_TASK_TIMEOUT");
        int timeout = (timeoutEnv != null && !timeoutEnv.isEmpty())
                          ? Integer.parseInt(timeoutEnv)
                          : 300;

        // Start async task execution
        ExecutionResult result = agent.getBrowser().executeTask(task, false, OutputSchema.class);

        assertTrue("Async task execution should succeed", result.isSuccess());
        assertNotNull("Request ID should not be null", result.getRequestId());
        assertFalse("Request ID should not be empty",
                    result.getRequestId().isEmpty());
        assertEquals("Error message should be empty", "",
                     result.getErrorMessage());
        assertNotNull("Task ID should not be null", result.getTaskId());
        assertFalse("Task ID should not be empty",
                    result.getTaskId().isEmpty());

        // Poll task status until finished
        int pollInterval = 3;
        int maxPollAttempts = timeout / pollInterval;
        int retryTimes = 0;
        QueryResult queryResult = null;

        while (retryTimes < maxPollAttempts) {
          queryResult = agent.getBrowser().getTaskStatus(result.getTaskId());

          assertTrue("Query should succeed", queryResult.isSuccess());
          System.out.println("⏳ Task " + queryResult.getTaskId() +
                             " running 🚀: " + queryResult.getTaskAction());

          if ("finished".equals(queryResult.getTaskStatus())) {
            break;
          }

          retryTimes++;
          Thread.sleep(3000);
        }

        // Verify the final task status
        assertNotNull("Query result should not be null", queryResult);
        assertTrue("Task should finish within timeout",
                   retryTimes < maxPollAttempts);
        assertEquals("Task status should be finished", "finished",
                     queryResult.getTaskStatus());

        System.out.println("✅ Task result: " + queryResult.getTaskProduct());
      }
    }
    /**
     * Main method to run tests manually (for debugging purposes).
     * In production, use Maven or IDE test runners.
     */
    public static void main(String[] args) {
        System.out.println("=== Running Agent Integration Tests ===\n");
        
        computer = new TestComputerAgent();
        browser = new TestBrowserAgent();
        try {
            
            // Run testExecuteTaskSuccess (currently skipped)
            System.out.println("\n--- Test 2: Computer Execute Task Success ---");
            computer.setUp();
            computer.testExecuteTaskSuccess();
            computer.tearDown();
            
            // Run testAsyncExecuteTaskSuccess (currently skipped)
            System.out.println("\n--- Test 3: Computer Async Execute Task Success ---");
            computer.setUp();
            computer.testAsyncExecuteTaskSuccess();
            computer.tearDown();

            System.out.println("\n--- Test 4: Browser Async Execute Task Success ---");
            computer.setUp();
            computer.testExecuteTaskSuccess();
            computer.tearDown();

            System.out.println("\n--- Test 5: Browser Async Execute Task Success ---");
            computer.setUp();
            computer.testAsyncExecuteTaskSuccess();
            computer.tearDown();
            System.out.println("\n=== All Tests Completed ===");
            System.out.println("\nNote: Some tests are currently skipped because Agent task execution");
            System.out.println("      methods are not yet implemented in the Java SDK.");
            System.out.println("      Once implemented, uncomment the test code to enable full testing.");
            
        } catch (Exception e) {
            System.err.println("❌ Test failed: " + e.getMessage());
            e.printStackTrace();
        }
    }
}

